src/plugins/ucc_discovery_plugin.cpp

```diff
@@ -180,57 +213,101 @@ bool UccDiscoveryPlugin::_dictionary_segments_contain_duplicates(const std::shar
-          if (distinct_value_count + 1 != distinct_values.size()) {
-            break;
-          }
-          ++distinct_value_count;
-          ++it;
-        }
-      });
+    if (source_chunk->invalid_row_count() == 0 && !source_chunk->is_mutable()) {
+      // The set of distinct values across all segments should grow by the number of rows in the segment because,
+      // if it does not, it means some value must have been inserted twice -> duplicate detected.
+      // In this case, the UCC is violated.
+      const auto expected_distinct_value_count = distinct_values_across_segments.size() + source_segment->size();
+
+      // If we enter this branch, we know that this segment has not been modified since its creation,
+      // so there is no need to use the validate operator indirection to access its values.
+      // We can do so directly on the segment instead.
+      if (const auto& value_segment = std::dynamic_pointer_cast<ValueSegment<ColumnDataType>>(source_segment)) {
+        // Directly insert all values.
+        const auto& values = value_segment->values();
+        distinct_values_across_segments.insert(values.cbegin(), values.cend());
+      } else if (const auto& dictionary_segment =
+                     std::dynamic_pointer_cast<DictionarySegment<ColumnDataType>>(source_segment)) {
+        // Directly insert dictionary entries.
+        const auto& dictionary = dictionary_segment->dictionary();
+        distinct_values_across_segments.insert(dictionary->cbegin(), dictionary->cend());
+      } else {
+        // We will check this segment later when we do the transaction-safe access to potentially modified chunks.
+        continue;
+      }
+
+      if (distinct_values_across_segments.size() != expected_distinct_value_count) {
+        return false;
+      }
+
+      // If we managed to check this segment already, we don't need to do it again later using the validate operator.
+      unmodified_chunks.push_back(chunk_id);
     }
+  }
+
+  // Using the validate operator, get a view of the current content of the table, filtering out overwritten and deleted
+  // values.
+  const auto logical_table = std::make_shared<GetTable>(table_name, unmodified_chunks, std::vector<ColumnID>());
+  logical_table->execute();
+  const auto validate_table_operator = std::make_shared<Validate>(logical_table);
+  validate_table_operator->set_transaction_context(transaction_context);
+  validate_table_operator->execute();
+  const auto& table_view = validate_table_operator->get_output();
+
+  // Check all chunks for duplicates that we haven't yet in the first loop above.
+  // Note that the loop below will only contain these chunks, as we have excluded the others when executing the GetTable
+  // operator.
+  const auto validated_chunk_count = table_view->chunk_count();
+  for (auto chunk_id = ChunkID{0}; chunk_id < validated_chunk_count; ++chunk_id) {
+    // No need to do null checks here as we already did so above
```

Comment:
Ah, you mean chunk nullptr and not NULL values. There's no need to check after the GetTable operator (it only forwards chunks with actual data).